### PR TITLE
Honour policyPeerShareActivationDelay timeout

### PR DESCRIPTION
**Pull request by karknu**
# Description
The policyPeerShareActivationDelay was effectivly ignored.
Commit 194ea32b removed the timeout logic because it was broken. This
was because availableForPeerShare contained a mix of peers that supported
and didn't support peersharing. As soon as a peer that doesn't support
peersharing is more than 900s old it would prevent any timeouts from
taking effect.
    
This commit change it so that only peers that support peersharing are
added to nextPeerShareTimes/availableForPeerShare. With that the
Guarskip related minPeerShareTime can be restored.

# Checklist

- Branch
    - [x] Updated changelog files.
    - [x] Commit sequence broadly makes sense
    - [x] Commits have useful messages
    - [ ] The documentation has been properly updated
    - [ ] New tests are added if needed and existing tests are updated
    - [ ] If serialization changes, user-facing consequences (e.g. replay from genesis) are confirmed to be intentional.
- Pull Request
    - [x] Self-reviewed the diff
    - [x] Useful pull request description at least containing the following information:
      - What does this PR change?
      - Why these changes were needed?
      - How does this affect downstream repositories and/or end-users?
      - Which ticket does this PR close (if any)? If it does, is it [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)?
    - [ ] Reviewer requested

### CODE REVIEW

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/KnownPeers.hs [hunk 1]
```diff
@@ -78,18 +78,14 @@ belowTarget actions
     -- We can only ask ones where we have not asked them within a certain time.
   , not (Set.null availableForPeerShare)
 
-    -- Are there any peers from the availableForPeerShare set we can really ask
-    -- (i.e. have correct PeerSharing permissions)?
-  , not (Set.null canAsk)
-
```
Comment (bolt12) [inline ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/KnownPeers.hs:84]: Why was this removed? the `availableForPeerShare` set contains peers which might have `PeerSharingDisabled` so we can't issue requests to them.

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/KnownPeers.hs [hunk 2]
```diff
@@ -78,18 +78,14 @@ belowTarget actions
...
     -- No peer share requests should be issued when the node is in a sensitive
     -- state
   , not (requiresBootstrapPeers bootstrapPeersFlag ledgerStateJudgement)
   = Guarded Nothing $ do
       -- Max selected should be <= numPeerShareReqsPossible
       selectedForPeerShare <- pickPeers st
                               policyPickKnownPeersForPeerShare
-                              canAsk
+                              availableForPeerShare
```
Comment (bolt12) [inline ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/KnownPeers.hs:88]: This might pick peers which don't have peer sharing enabled resulting in a protocol failure

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/State/EstablishedPeers.hs [hunk 3]
```diff
@@ -97,10 +97,11 @@ invariant :: Ord peeraddr
...
           -> Bool
 invariant EstablishedPeers {..} =
-       -- The combo of the peer share set + psq = the whole set of peers
-       availableForPeerShare
+       -- The combo of the peer share set + psq = a subset of all peers
+ (  availableForPeerShare
     <> Set.fromList (PSQ.keys nextPeerShareTimes)
-    == Map.keysSet allPeers
+ )
+    `Set.isSubsetOf` Map.keysSet allPeers
```
Comment (bolt12) [inline ouroboros-network/src/Ouroboros/Network/PeerSelection/State/EstablishedPeers.hs:104]: I don't like the fact that this was weakened. Why do we need it?

**Review by bolt12 — CHANGES_REQUESTED** [review]

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/State/EstablishedPeers.hs [hunk 4]
```diff
@@ -162,13 +163,16 @@ member peeraddr = Map.member peeraddr . allPeers
...
-insert peeraddr peerconn peerShareAt ep@EstablishedPeers { allPeers } =
+insert peeraddr peerconn peerShareAt_m ep@EstablishedPeers { allPeers } =
    -- Ask newly established peers for its peers after the specified delay.
-   setPeerShareTime (Set.singleton peeraddr) peerShareAt $
-     ep { allPeers = Map.insert peeraddr peerconn allPeers }
+   let ep' = ep { allPeers = Map.insert peeraddr peerconn allPeers } in
+   case peerShareAt_m of
+        Nothing          -> ep'
+        Just peerShareAt ->
+          setPeerShareTime (Set.singleton peeraddr) peerShareAt $ ep'
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/PeerSelection/State/EstablishedPeers.hs:175]: I think this is a bit cleaner (at the expense of code duplication, which could be eliminated with a recursive call):
```suggestion
insert peeraddr peerconn Nothing ep@EstablishedPeers { allPeers } =
    ep { allPeers = Map.insert peeraddr peerconn allPeers }   
insert peeraddr peerconn (Just peerShareAt) ep@EstablishedPeers { allPeers } =
    -- Ask newly established peers for its peers after the specified delay.
    setPeerShareTime (Set.singleton peeraddr) $
    ep { allPeers = Map.insert peeraddr peerconn allPeers } 
```

**Review by coot — APPROVED** [review]
LGTM, just a minor comment.

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/KnownPeers.hs (hunk 1 above)
Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/KnownPeers.hs:84]: As I understand it it cannot, because we only add peers which enabled peer sharing to `availableForPeerShare`.

Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/KnownPeers.hs:84]: I changed it so that `availableForPeerShare` only contains peers that actually support peersharing. This removes the need for `canAsk`.

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/State/EstablishedPeers.hs [hunk 5]
```diff
@@ -162,13 +163,16 @@ member peeraddr = Map.member peeraddr . allPeers
 insert :: Ord peeraddr
        => peeraddr
        -> peerconn
-       -> Time
+       -> Maybe Time
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/PeerSelection/State/EstablishedPeers.hs:166]: Could you add a haddock; it's important to use this argument right.

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/KnownPeers.hs (hunk 2 above)
Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/KnownPeers.hs:88]: No, see previous comment regarding `availableForPeerShare`.

File: ouroboros-network/sim-tests-lib/Test/Ouroboros/Network/Testnet.hs [hunk 6]
```diff
@@ -837,7 +837,7 @@ prop_track_coolingToCold_demotions defaultBearerInfo diffScript =
           notInProgressDemoteToColdForTooLong =
             map (\addr ->
                   Signal.keyedTimeoutTruncated
-                    60
+                    120
```
Comment (coot) [inline ouroboros-network/sim-tests-lib/Test/Ouroboros/Network/Testnet.hs:840]: Could you explain why we need to increase it?

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/State/EstablishedPeers.hs (hunk 3 above)
Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/PeerSelection/State/EstablishedPeers.hs:104]: It is needed because `allPeers` contains both peers that support and doesn't support peersharing, but `availableForPeerShare` and `nextPeerShareTimes` only contains peers that support peersharing.

File: ouroboros-network/sim-tests-lib/Test/Ouroboros/Network/Testnet.hs (hunk 6 above)
Comment (bolt12) [reply ouroboros-network/sim-tests-lib/Test/Ouroboros/Network/Testnet.hs:840]: There was a test failure. I looked into it and it was due to the timeout just barely making it.

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/State/EstablishedPeers.hs (hunk 3 above)
Comment (bolt12) [reply ouroboros-network/src/Ouroboros/Network/PeerSelection/State/EstablishedPeers.hs:104]: I would add a clause to the invariant that checks that all peers in the `availableForPeerShare` set have peer sharing enabled then.

File: ouroboros-network/sim-tests-lib/Test/Ouroboros/Network/Testnet.hs (hunk 6 above)
Comment (bolt12) [reply ouroboros-network/sim-tests-lib/Test/Ouroboros/Network/Testnet.hs:840]: It isn't

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/State/EstablishedPeers.hs (hunk 5 above)
Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/PeerSelection/State/EstablishedPeers.hs:166]: :+1: Will do.

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/State/EstablishedPeers.hs (hunk 3 above)
Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/PeerSelection/State/EstablishedPeers.hs:105]: I've added a check which makes sure that only peers which support peersharing are in `availableForPeerShare <> nextPeerShareTimes` .

**Review by bolt12 — APPROVED** [review]
